src/pint/models/fdjump.py

```diff
@@ -0,0 +1,207 @@
+
+import re
+from warnings import warn
+
+import astropy.units as u
+import numpy as np
+
+from pint.models.parameter import boolParameter, maskParameter
+from pint.models.timing_model import DelayComponent
+
+fdjump_max_index = 20
+
+
+class FDJump(DelayComponent):
+    """A timing model for system-dependent frequency evolution of pulsar
+    profiles.
+
+    This model expresses the delay as a polynomial function of the
+    observing frequency/logarithm of observing frequency in the SSB frame.
+    This is intended to compensate for the delays introduced by frequency-dependent
+    profile structure when a different profiles are used for different systems.
+
+    The default behavior is to have FDJUMPs as polynomials of the observing
+    frequency (rather than log-frequency). This is different from the convention
+    used for global FD parameters. This choice is made to be compatible with tempo2.
+    This is controlled using the `FDJUMPLOG` parameter. `FDJUMPLOG Y` may not be
+    tempo2-compatible.
+
+    Caveat
+    ------
+    `FDJUMP`s have two indices: the polynomial/FD/prefix index and the system/mask
+    index. i.e., they have properties of both `maskParameter`s such as `JUMP`s and
+    `prefixParameters` such as `FD`s. There is currently no elegant way in PINT
+    to implement such parameters due to the way parameter indexing is implemented;
+    there is no way to distinguish between mask and prefix indices.
+
+    Hence, they are implemented here as `maskParameter`s as a stopgap measure.
+    This means that there must be an upper limit for the FD indices. This is controlled
+    using the ``pint.models.fdjump.fdjump_max_index`` variable, and is 20 by default.
+    Note that this is strictly a limitation of the implementation and not a property
+    of `FDJUMP`s themselves.
+
+    `FDJUMP`s appear in tempo2-format par files as `FDJUMPp`, where p is the FD index.
+    The mask index is not explicitly mentioned in par files similar to `JUMP`s.
+    PINT understands both `FDJUMPp` and `FDpJUMP` as the same parameter in par files,
+    but the internal representation is always `FDpJUMPq`, where q is the mask index.
+
+    PINT understands 'q' as the mask parameter just fine, but the identification of 'p'
+    as the prefix parameter is done in a hacky way.
+
+    This implementation may be overhauled in the future.
+
+    Parameters supported:
+
+    .. paramtable::
+        :class: pint.models.fdjump.FDJump
+    """
+
+    register = True
+    category = "fdjump"
+
+    def __init__(self):
+        super().__init__()
+
+        # Matches "FDpJUMPq" where p and q are integers.
+        self.param_regex = re.compile("^FD(\\d+)JUMP(\\d+)")
+
+        self.add_param(
+            boolParameter(
+                name="FDJUMPLOG",
+                value=False,
+                description="Whether to use log-frequency for computing FDJUMPs.",
+            )
+        )
+        for j in range(1, fdjump_max_index + 1):
```

Comment:
So when you add this component, do jumps 1-20 all get added right away?

Reply:
If so, you might explain that too in the docs, and how to keep the unneeded ones from having any effect.